src/ir/child-typer.h

```diff
@@ -815,8 +825,8 @@ template<typename Subtype> struct ChildTyper : OverriddenVisitor<Subtype> {
   }
 
   void visitTupleMake(TupleMake* curr) {
-    for (auto& expr : curr->operands) {
-      noteAny(&expr);
+    for (Index i = 0; i < curr->operands.size(); ++i) {
+      note(&curr->operands[i], VarType{Index(curr->operands.size() - i - 1)});
```

Comment:
Why is this not
```suggestion
      note(&curr->operands[i], VarType{Index(i)});
```
?

Reply:
Because of the rules about the order in which variables are introduced in `PrincipalType`s. The variables have to be introduced in order with zero at the end. Notably we're not actually using `PrincipalType`s here, but this will eventually be used to construct `PrincipalType`s.

Reply:
What does "in order" mean here? I'm not sure what order this is matching (the spec, or `principal-types.h`, or something else; I do see that `principal-types.h` mentions that params are stored in reverse order, but (1) that is an internal detail, and (2) you mentioned variables, not params specifically, so I'm not sure if it's related or not)

Reply:
From principal-type.h:
```
// ... The parameters are stored in reverse order
// for efficiency. Variable indices for each kind of variable must be introduced
// contiguously in order starting at zero from the end to beginning of the
// params (beginning to end of the reversed params).
```

Reply:
Oh, thanks, I missed that. So while the user provides the params in order, the indices must be in reverse order... perhaps the user should provide the params already reversed, as this is observable anyhow, and not just an internal detail? (But that would be separate from this PR)

Reply:
Oh, you were talking about when the params are provided to the `PrincipalType` constructor. Yes, that's certainly on the table as a potential simplification, too.